gitlab/resource_gitlab_project_variable.go

```diff
@@ -78,80 +108,162 @@ func resourceGitlabProjectVariableCreate(d *schema.ResourceData, meta interface{
+		var variableType gitlab.VariableTypeValue
+		var protected bool
+		var masked bool
+		if environmentScope == environmentScopeOfProjectVariableBeingUpdated {
+			value = d.Get("value").(string)
+			variableType = *stringToVariableType(d.Get("variable_type").(string))
+			protected = d.Get("protected").(bool)
+			masked = d.Get("masked").(bool)
+		} else {
+			value = projectVariable.Value
+			variableType = projectVariable.VariableType
+			protected = projectVariable.Protected
+			masked = projectVariable.Masked
+		}
+		options := gitlab.CreateProjectVariableOptions{
+			Key:              &key,
+			Value:            &value,
+			VariableType:     &variableType,
+			Protected:        &protected,
+			Masked:           &masked,
+			EnvironmentScope: &environmentScope,
+		}
+		log.Printf("[DEBUG] create gitlab project variable %s/%s/%s", project, key, environmentScope)
+		_, _, err := client.ProjectVariables.CreateVariable(project, &options)
+		if err != nil {
+			return err
+		}
+	}
 	return resourceGitlabProjectVariableRead(d, meta)
 }
 
 func resourceGitlabProjectVariableDelete(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*gitlab.Client)
-	project := d.Get("project").(string)
+	project := d.Get("project")
 	key := d.Get("key").(string)
-	log.Printf("[DEBUG] Delete gitlab project variable %s/%s", project, key)
+	environmentScopeOfProjectVariableBeingDeleted := d.Get("environment_scope")
+
+	projectVariablesWithSameKeyAsTheOneBeingDeleted, fetchErr := projectVariablesMatchingKey(client, project, key)
+	if fetchErr != nil {
+		return fetchErr
+	}
+
+	deleteErr := deleteProjectVariables(projectVariablesWithSameKeyAsTheOneBeingDeleted, project, client)
+	if deleteErr != nil {
+		return deleteErr
+	}
+
+	for _, projectVariable := range projectVariablesWithSameKeyAsTheOneBeingDeleted {
+		if projectVariable.EnvironmentScope == environmentScopeOfProjectVariableBeingDeleted {
+			continue // Don't re-create project variable for the environment scope that is marked for deletion
+		}
+		options := gitlab.CreateProjectVariableOptions{
+			Key:              &projectVariable.Key,
+			Value:            &projectVariable.Value,
+			VariableType:     &projectVariable.VariableType,
+			Protected:        &projectVariable.Protected,
+			Masked:           &projectVariable.Masked,
+			EnvironmentScope: &projectVariable.EnvironmentScope,
+		}
+		log.Printf("[DEBUG] create gitlab project variable %s/%s/%s", project, key, projectVariable.EnvironmentScope)
+		_, _, err := client.ProjectVariables.CreateVariable(project, &options)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func deleteProjectVariables(projectVariables []*gitlab.ProjectVariable, project interface{}, client *gitlab.Client) error {
+	for _, projectVariable := range projectVariables {
+		log.Printf("[DEBUG] Delete gitlab project variable %s/%s", project, projectVariable.Key)
+		_, err := client.ProjectVariables.RemoveVariable(project, projectVariable.Key)
```

Comment:
the gitlab api still has no way to deterministically target a variable for modification or deletion, even if you're able to do an accurate lookup. i don't believe the EnvironmentScope option in this context is used by the API to identify a variable. Rather, the API seems to use the variable Key exclusively and the additional options are just configuring the variable's properties. 

My belief is that if this call will just delete a variable with the specified key but and a random environment scope.

Reply:
We are aligned in our understanding of the Gitlab API behavior.

Hence, the code applies a brute force approach here.

1. Fetch all ProjectVariables
2. Filter ProjectVariables for the key that is being updated or deleted
3. Issue n Delete api calls for ProjectVariable key
4. Re-create without change ProjectVariables for key & environmentScope that are not being updated/deleted
5. If Update, re-create with change ProjectVariable for key & environmentScope that is being updated
5. If Delete, no API call for the ProjectVariable for key & environmentScope that is should be deleted

Reply:
Clever workaround. Might be dicey touching variables potentially not managed by terraform or managed by a different project. But as a workaround I agree it's better than nothing at this point.